python/ray/serve/backend_state_manager.py

```diff
@@ -977,15 +1082,22 @@ def update(self) -> bool:
             slow_start_replicas = []
             for replica in replicas.pop(
                     states=[ReplicaState.STARTING_OR_UPDATING]):
-                ready, slow_start = replica.check_started()
-                if ready:
+                state = replica.check_started()
+                if state == ReplicaState.RUNNING:
                     # This replica should be now be added to handle's replica
                     # set.
                     replicas.add(ReplicaState.RUNNING, replica)
                     transitioned_backend_tags.add(backend_tag)
+                elif state == ReplicaState.FAILED:
+                    # Replica reconfigure (deploy / upgrade) failed after all
+                    # retries
+                    replica.set_should_stop(0)
+                    replicas.add(ReplicaState.FAILED, replica)
```

Comment:
Hmm it looks like we never remove replicas from the `FAILED` state. I think we can remove this in two cases: the deployment fully failed (I guess we nuke the deployment in that case anyways) or the deployment succeeded.